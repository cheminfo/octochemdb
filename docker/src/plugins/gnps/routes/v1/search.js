import { searchHandler } from './searchHandler/searchHanlder.js';

// export the handler
const searchGNPS = {
  method: ['GET', 'POST'],
  schema: {
    summary: 'Retrieve mass spectra from range of masses, from EM or from MF',
    description:
      'Allows to search for mass spectra based on a range of m/z values. Also allows to search for exact mass or molecular formula.',
    querystring: {
      em: {
        type: 'string',
        description: 'Monoisotopic mass (in Da)',
        example: '980.533, 939.451',
        default: '',
      },
      masses: {
        type: 'string',
        description: 'List experimental mass',
        example: '289.2,295.5',
        default: '',
      },
      precision: {
        type: 'number',
        description: 'Precision (in ppm) of the monoisotopic mass',
        default: 10,
      },
      limit: {
        type: 'number',
        description: 'Maximum number of results to return',
        default: 10,
      },
      mf: {
        type: 'string',
        description: 'MF of the compound',
        example: 'C48H72N10O12',
        default: '',
      },
      fields: {
        type: 'string',
        description: 'Fields to retrieve',
        default: 'data.spectrum,data.ocl',
      },
    },
  },
  handler: searchHandler,
};
export default searchGNPS;
