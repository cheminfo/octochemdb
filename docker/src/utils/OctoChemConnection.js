import delay from 'delay';
import dotenv from 'dotenv';
import md5 from 'md5';
import { MongoClient } from 'mongodb';

import debugLibrary from './Debug.js';

dotenv.config();
const debug = debugLibrary('OctoChemConnection');

export function OctoChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

OctoChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

OctoChemConnection.prototype.getCollectionNames =
  async function getCollectionNames() {
    return (await (await this.getDatabase()).listCollections().toArray()).map(
      (entry) => entry.name,
    );
  };

OctoChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

OctoChemConnection.prototype.getImportationLog =
  async function getImportationLog(options) {
    const { collectionName, sources, startSequenceID } = options;
    const logsCollection = await this.getImportationLogsCollection();
    const sourcesHash = md5(JSON.stringify(sources));
    const _id = sourcesHash;
    let logs = await logsCollection.find({ _id }).next();
    if (logs === null) {
      logs = {
        _id,
        collectionName,
        sources,
        sourcesHash,
        dateStart: Date.now(),
        dateEnd: Date.now(),
        startSequenceID,
        endSequenceID: 0,
        status: 'updating',
      };
      await logsCollection.insertOne(logs);
    }
    return logs;
  };
OctoChemConnection.prototype.updateImportationLog =
  async function updateImportationLog(logs) {
    const collection = await this.getImportationLogsCollection();
    logs.dateEnd = Date.now();
    await collection.replaceOne({ _id: logs._id }, logs);
  };

OctoChemConnection.prototype.getImportationLogsCollection =
  async function getCollection() {
    return (await this.getDatabase()).collection('importationLogs');
  };

OctoChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

OctoChemConnection.prototype.setProgress = async function setProgress(
  progress,
) {
  const collection = await this.getAdminCollection();
  await collection.replaceOne({ _id: progress._id }, progress);
};

OctoChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    progress = {
      _id,
      state: 'updating',
      seq: 0,
      dateStart: Date.now(),
      dateEnd: 0,
    };
    await adminCollection.insertOne(progress);
  }
  return progress;
};

OctoChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      debug(e.stack);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

OctoChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};
