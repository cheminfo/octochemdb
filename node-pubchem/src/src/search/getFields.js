import Debug from 'debug';

const debug = Debug('getFields');

function getFields(fields = '') {
  let result = { id: '$_id' };
  for (let field of fields.split(',').filter((field) => field)) {
    result[field] = 1;
  }
  debug(`List of required fields: ${JSON.stringify(result)}`);
  return result;
}

export default getFields;
