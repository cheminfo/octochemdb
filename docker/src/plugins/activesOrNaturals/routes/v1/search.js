import escapeRegExp from 'lodash.escaperegexp';
import { MF } from 'mf-parser';

import { getFields, OctoChemConnection } from '../../../../server/utils.js';
import debugLibrary from '../../../../utils/Debug.js';

const debug = debugLibrary('entriesSearch');
// export the handler
const entriesSearch = {
  method: ['GET', 'POST'],
  schema: {
    summary: 'Retrieve compounds from a monoisotopic mass',
    description:
      'Allows to search for compounds based on a monoisotopic mass and precision (accuracy) of the measurement. \n Optional parameters can be used to filter the results based on their taxonomy, bioactivity and topic (MeSH terms) of PubMed publications related to the the molecules.',
    querystring: {
      em: {
        type: 'string',
        description: 'Monoisotopic mass (in Da)',
        example: '300.123, 259.0237',
        default: '',
      },
      precision: {
        type: 'number',
        description: 'Precision (in ppm) of the monoisotopic mass',
        default: 100,
      },
      mf: {
        type: 'string',
        description: 'MF of the compound',
        example: '',
      },
      kwActiveAgainst: {
        type: 'string',
        description:
          'Taxonomies superkingdom, kingdom or phylum of target organism in bioassays(separate terms to search with ";" or "," )',
        example: 'Halobacterium salinarum',
        default: '',
      },
      kwTaxonomies: {
        type: 'string',
        description:
          'Taxonomies family, genus or species (separate terms to search with ";" or "," )',
        example: 'Podocarpus macrophyllus',
        default: '',
      },
      kwBioassays: {
        type: 'string',
        description: 'keywords bioassays',
        example: 'MIC',
        default: '',
      },
      kwMeshTerms: {
        type: 'string',
        description: 'keywords mesh terms (separate terms to search with ";" )',
        example: 'antibiotic',
        default: '',
      },
      isNaturalProduct: {
        type: 'boolean',
        description:
          'if true, only natural products are returned, if false, only not natural compounds are returned and if undefined, both are returned',
        default: undefined,
      },
      isBioactive: {
        type: 'boolean',
        description:
          'if true, only bioactive compounds are returned, if false, only not bioactive compounds are returned and if undefined, both are returned',
        default: undefined,
      },
      minNbMassSpectra: {
        type: 'number',
        description: 'minimum number of mass spectra',
        default: undefined,
      },
      maxNbMassSpectra: {
        type: 'number',
        description: 'Maximum number of mass spectra',
        default: undefined,
      },
      minNbActivities: {
        type: 'number',
        description: 'Minimum number of activities',
        default: undefined,
      },
      maxNbActivities: {
        type: 'number',
        description: 'Maximum number of activities',
        default: undefined,
      },
      minNbTaxonomies: {
        type: 'number',
        description: 'Minimum number of taxonomies',
        default: undefined,
      },
      maxNbTaxonomies: {
        type: 'number',
        description: 'Maximum number of taxonomies',
        default: undefined,
      },
      minNbPatents: {
        type: 'number',
        description: 'Minimum number of patents',
        default: undefined,
      },
      maxNbPatents: {
        type: 'number',
        description: 'Maximum number of patents',
        default: undefined,
      },
      minNbPubmeds: {
        type: 'number',
        description: 'Minimum of PubMed publications',
        default: undefined,
      },
      maxNbPubmeds: {
        type: 'number',
        description: 'Maximum of PubMed publications',
        default: undefined,
      },
      limit: {
        type: 'number',
        description: 'Maximum number of results to return',
        default: 1000,
      },
      fields: {
        type: 'string',
        description: 'Fields to retrieve',
        default: 'data.em,data.mf',
      },
    },
  },
  handler: searchHandler,
};
export default entriesSearch;

/**
 * @description Search for compounds from a monoisotopic mass, target taxonomies, source taxonomies and bioassays
 * @param {object} request
 * @returns {Promise<object>} Entries who match the query parameters inside the activeOrNaturals collection
 */
async function searchHandler(request) {
  let data;
  if (request.method === 'GET') {
    data = request.query;
  } else {
    data = request.body;
  }
  let {
    em = '',
    mf = '',
    kwTaxonomies = '',
    kwBioassays = '',
    kwActiveAgainst = '',
    kwMeshTerms = '',
    isNaturalProduct = undefined,
    isBioactive = undefined,
    minNbMassSpectra = undefined,
    maxNbMassSpectra = undefined,
    minNbActivities = undefined,
    maxNbActivities = undefined,
    minNbTaxonomies = undefined,
    maxNbTaxonomies = undefined,
    minNbPatents = undefined,
    maxNbPatents = undefined,
    minNbPubmeds = undefined,
    maxNbPubmeds = undefined,
    limit = 1e3,
    precision = 100,
    fields = 'data.em,data.mf',
  } = data;
  // This keywords use regular expressions to search even for incomplete terms
  let wordsWithRegexBioassays = [];
  let wordsWithRegexMeshTerms = [];
  // convert to lower case and remove spaces and split by ";" or ","
  let wordsToBeSearchedBioassays = kwBioassays
    .toLowerCase()
    .split(/ *[,;\t\n\r\s]+ */)
    .filter((entry) => entry);

  let wordsToBeSearchedActiveAgainst = kwActiveAgainst
    .toLowerCase()
    .split(/ *[,;\t\n\r\s]+ */)
    .filter((entry) => entry);

  let wordsToBeSearchedTaxonomies = kwTaxonomies
    .toLowerCase()
    .split(/ *[,;\t\n\r\s]+ */)
    .filter((entry) => entry);

  let wordsToBeSearchedMeshTerms = kwMeshTerms
    .toLowerCase()
    .split(/ *[,;\t\n\r\s]+ */)
    .filter((entry) => entry);
  // convert words to be searched in bioassays to regex
  wordsToBeSearchedBioassays.forEach((word) => {
    wordsWithRegexBioassays.push(new RegExp(escapeRegExp(word), 'i'));
  });
  // convert phrases to regular expressions
  wordsToBeSearchedMeshTerms.forEach((word) => {
    wordsWithRegexMeshTerms.push(new RegExp(escapeRegExp(word), 'i'));
  });

  // define lower and upper bounds of the returned results limit
  if (limit > 1e4) limit = 1e4;
  if (limit < 1) limit = 1;
  // define the error allowed for the search

  let connection;
  try {
    connection = new OctoChemConnection();
    const collection = await connection.getCollection('activesOrNaturals');
    // get the fields to be retrieved
    let formattedFields = getFields(fields);
    // define match parameters for the search, the $in operator is used to search for multiple words and is true if at least one of the words is found
    let matchParameter = {};

    if (mf !== '') {
      let mfinfo = new MF(mf).getInfo();
      matchParameter['data.mf'] = mfinfo.mf;
    }
    let error;
    let ems = em
      .split(/[ ,;\t\r\n]+/)
      .filter((entry) => entry)
      .map(Number);
    if (ems.length > 1) {
      let match = [];

      for (let em of ems) {
        error = (em / 1e6) * precision;
        match.push({
          'data.em': { $lt: em + error, $gt: em - error },
        });
      }
      matchParameter = { $or: match };
    } else if (ems.length === 1 && ems[0] !== '') {
      error = (ems[0] / 1e6) * precision;

      matchParameter = {
        'data.em': { $lt: ems[0] + error, $gt: ems[0] - error },
      };
    }
    if (kwTaxonomies) {
      matchParameter['data.kwTaxonomies'] = {
        $in: wordsToBeSearchedTaxonomies,
      };
    }
    if (kwBioassays) {
      matchParameter['data.kwBioassays'] = { $in: wordsWithRegexBioassays };
    }
    if (kwMeshTerms) {
      matchParameter['data.kwMeshTerms'] = { $in: wordsWithRegexMeshTerms };
    }
    if (kwActiveAgainst) {
      matchParameter['data.kwActiveAgainst'] = {
        $in: wordsToBeSearchedActiveAgainst,
      };
    }
    if (isNaturalProduct !== undefined) {
      matchParameter['data.naturalProduct'] = isNaturalProduct;
    }
    if (isBioactive !== undefined) {
      matchParameter['data.bioactive'] = isBioactive;
    }
    if (minNbMassSpectra !== undefined && maxNbMassSpectra !== undefined) {
      matchParameter['data.nbMassSpectra'] = {
        $gte: minNbMassSpectra,
        $lte: maxNbMassSpectra,
      };
    } else if (
      minNbMassSpectra !== undefined &&
      maxNbMassSpectra === undefined
    ) {
      matchParameter['data.nbMassSpectra'] = { $gte: minNbMassSpectra };
    } else if (
      maxNbMassSpectra !== undefined &&
      minNbMassSpectra === undefined
    ) {
      matchParameter['data.nbMassSpectra'] = { $lte: maxNbMassSpectra };
    }
    if (minNbActivities !== undefined && maxNbActivities !== undefined) {
      matchParameter['data.nbActivities'] = {
        $gte: minNbActivities,
        $lte: maxNbActivities,
      };
    } else if (minNbActivities !== undefined && maxNbActivities === undefined) {
      matchParameter['data.nbActivities'] = { $gte: minNbActivities };
    } else if (maxNbActivities !== undefined && minNbActivities === undefined) {
      matchParameter['data.nbActivities'] = { $lte: maxNbActivities };
    }
    if (minNbTaxonomies !== undefined && maxNbTaxonomies !== undefined) {
      matchParameter['data.nbTaxonomies'] = {
        $gte: minNbTaxonomies,
        $lte: maxNbTaxonomies,
      };
    } else if (minNbTaxonomies !== undefined && maxNbTaxonomies === undefined) {
      matchParameter['data.nbTaxonomies'] = { $gte: minNbTaxonomies };
    } else if (maxNbTaxonomies !== undefined && minNbTaxonomies === undefined) {
      matchParameter['data.nbTaxonomies'] = { $lte: maxNbTaxonomies };
    }
    if (minNbPatents !== undefined && maxNbPatents !== undefined) {
      matchParameter['data.nbPatents'] = {
        $gte: minNbPatents,
        $lte: maxNbPatents,
      };
    } else if (minNbPatents !== undefined && maxNbPatents === undefined) {
      matchParameter['data.nbPatents'] = { $gte: minNbPatents };
    } else if (maxNbPatents !== undefined && minNbPatents === undefined) {
      matchParameter['data.nbPatents'] = { $lte: maxNbPatents };
    }
    if (minNbPubmeds !== undefined && maxNbPubmeds !== undefined) {
      matchParameter['data.nbPubmeds'] = {
        $gte: minNbPubmeds,
        $lte: maxNbPubmeds,
      };
    } else if (minNbPubmeds !== undefined && maxNbPubmeds === undefined) {
      matchParameter['data.nbPubmeds'] = { $gte: minNbPubmeds };
    } else if (maxNbPubmeds !== undefined && minNbPubmeds === undefined) {
      matchParameter['data.nbPubmeds'] = { $lte: maxNbPubmeds };
    }

    // search for the entries
    const results = await collection
      .aggregate([
        { $match: matchParameter },
        { $limit: limit },
        {
          $project: formattedFields,
        },
      ])
      .toArray();
    return { data: results };
  } catch (e) {
    if (connection) {
      await debug.fatal(e.message, {
        collection: 'activesOrNaturals',
        connection,
        stack: e.stack,
      });
    }
    return { errors: [{ title: e.message, detail: e.stack }] };
  } finally {
    debug.trace('Closing connection');
    if (connection) await connection.close();
  }
}
