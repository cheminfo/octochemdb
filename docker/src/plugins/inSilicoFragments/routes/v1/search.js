import { getFields, OctoChemConnection } from '../../../../server/utils.js';
import debugLibrary from '../../../../utils/Debug.js';
import { getRequestQuery } from '../../../../utils/getRequestQuery.js';

const debug = debugLibrary('fromMasses');
// export the handler
const fromMasses = {
  method: ['GET', 'POST'],
  schema: {
    summary: 'Retrieve mass spectra from range of masses, from EM or from MF',
    description:
      'Allows to search for mass spectra based on a range of m/z values. Also allows to search for exact mass or molecular formula.',
    querystring: {
      masses: {
        type: 'string',
        description: 'List experimental mass',
        example: '125.02,185.09',
        default: '',
      },
      precision: {
        type: 'number',
        description: 'Precision (in ppm) of the monoisotopic mass',
        default: 10,
      },
      mode: {
        type: 'string',
        description: 'Mode of the mass spectra: positive or negative',
        default: 'positive',
      },
      limit: {
        type: 'number',
        description: 'Maximum number of results to return',
        default: 10,
      },
      fields: {
        type: 'string',
        description: 'Fields to retrieve',
        default: 'data.masses.positive,data.ocl',
      },
    },
  },
  handler: searchHandler,
};
export default fromMasses;

/**
 * @description Search for compounds from a monoisotopic mass, target taxonomies, source taxonomies and bioassays
 * @param {object} request
 * @returns {Promise<object>} Entries who match the query parameters inside the activeOrNaturals collection
 */
async function searchHandler(request) {
  let data = getRequestQuery(request);
  let {
    masses = '',
    mode = 'positive',
    precision = 10,
    limit = 10,
    fields = 'data.masses,data.ocl',
  } = data;

  // define the error allowed for the search
  let connection;
  try {
    connection = new OctoChemConnection();
    const collection = await connection.getCollection('inSilicoFragments');
    // get the fields to be retrieved
    let formattedFields = getFields(fields);

    let massesArray = masses.split(/[, \t\n\r:;]+/);
    let matchParameter = {};
    let spectraPatamenters = [];
    let modeParameter;
    if (mode === 'positive') {
      modeParameter = 'data.masses.positive';
    } else {
      modeParameter = 'data.masses.negative';
    }
    if (massesArray.length > 0 && massesArray[0] !== '') {
      for (let massString of massesArray) {
        const mass = Number(massString);
        const error = (mass / 1e6) * precision;
        let search = {};
        search[modeParameter] = {
          $elemMatch: {
            $gte: mass - error,
            $lte: mass + error,
          },
        };

        spectraPatamenters.push(search);
      }
      matchParameter.$and = spectraPatamenters;
    }
    // search for the entries
    const results = await collection
      .aggregate([
        { $match: matchParameter },
        { $limit: limit },
        {
          $project: formattedFields,
        },
      ])
      .toArray();
    return { data: results };
  } catch (e) {
    if (connection) {
      await debug.error(e.message, {
        collection: 'inSilicoFragments',
        connection,
        stack: e.stack,
      });
    }
    return { errors: [{ title: e.message, detail: e.stack }] };
  } finally {
    if (connection) {
      debug.trace('Closing connection');
      await connection.close();
    }
  }
}
