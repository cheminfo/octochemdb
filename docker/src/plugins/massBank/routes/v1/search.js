import { MF } from 'mass-tools';

import { getFields, OctoChemConnection } from '../../../../server/utils.js';
import debugLibrary from '../../../../utils/Debug.js';
import { getRequestQuery } from '../../../../utils/getRequestQuery.js';

const debug = debugLibrary('fromMasses');
// export the handler
const fromMasses = {
  method: ['GET', 'POST'],
  schema: {
    summary: 'Retrieve mass spectra from range of masses, from EM or from MF',
    description:
      'Allows to search for mass spectra based on a range of m/z values. Also allows to search for exact mass or molecular formula.',
    querystring: {
      em: {
        type: 'string',
        description: 'Monoisotopic mass (in Da)',
        example: '287.116, 318.147',
        default: '',
      },
      masses: {
        type: 'string',
        description: 'List experimental mass',
        example: '125.02,185.09',
        default: '',
      },
      precision: {
        type: 'number',
        description: 'Precision (in ppm) of the monoisotopic mass',
        default: 10,
      },
      limit: {
        type: 'number',
        description: 'Maximum number of results to return',
        default: 10,
      },
      mf: {
        type: 'string',
        description: 'MF of the compound',
        example: 'C16H17NO4',
        default: '',
      },
      fields: {
        type: 'string',
        description: 'Fields to retrieve',
        default: 'data.spectrum,data.ocl',
      },
    },
  },
  handler: searchHandler,
};
export default fromMasses;

/**
 * @description Search for compounds from a monoisotopic mass, target taxonomies, source taxonomies and bioassays
 * @param {object} request
 * @returns {Promise<object>} Entries who match the query parameters inside the activeOrNaturals collection
 */
async function searchHandler(request) {
  let data = getRequestQuery(request);
  let {
    em = '',
    mf = '',
    masses = '',
    precision = 10,
    limit = 10,
    fields = 'data.spectrum,data.ocl',
  } = data;

  // define the error allowed for the search
  let connection;
  try {
    connection = new OctoChemConnection();
    const collection = await connection.getCollection('massBank');
    // get the fields to be retrieved
    let formattedFields = getFields(fields);

    let massesArray = masses.split(/[, \t\n\r:;]+/);
    let matchParameter = {};
    let spectraPatamenters = [];
    if (massesArray.length > 0 && massesArray[0] !== '') {
      for (let massString of massesArray) {
        const mass = Number(massString);
        const error = (mass / 1e6) * precision;
        spectraPatamenters.push({
          'data.spectrum.data.x': {
            $elemMatch: {
              $gte: mass - error,
              $lte: mass + error,
            },
          },
        });
      }
      matchParameter.$and = spectraPatamenters;
    }
    if (mf !== '') {
      let mfinfo = new MF(mf).getInfo();
      matchParameter['data.mf'] = mfinfo.mf;
    }
    let error;
    let ems = em
      .split(/[ ,;\t\r\n]+/)
      .filter((entry) => entry)
      .map(Number);
    if (ems.length > 1) {
      let match = [];

      for (let em of ems) {
        error = (em / 1e6) * precision;
        match.push({
          'data.em': { $lt: em + error, $gt: em - error },
        });
      }
      matchParameter = { $or: match };
    } else if (ems.length === 1 && ems[0] !== '') {
      error = (ems[0] / 1e6) * precision;

      matchParameter = {
        'data.em': { $lt: ems[0] + error, $gt: ems[0] - error },
      };
    }
    // search for the entries
    const results = await collection
      .aggregate([
        { $match: matchParameter },
        { $limit: limit },
        {
          $project: formattedFields,
        },
      ])
      .toArray();
    return { data: results };
  } catch (e) {
    if (connection) {
      await debug.error(e.message, {
        collection: 'massBank',
        connection,
        stack: e.stack,
      });
    }
    return { errors: [{ title: e.message, detail: e.stack }] };
  } finally {
    if (connection) {
      debug.trace('Closing connection');
      await connection.close();
    }
  }
}
