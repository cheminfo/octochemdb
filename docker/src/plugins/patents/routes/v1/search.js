import { OctoChemConnection, getFields } from '../../../../server/utils.js';
import debugLibrary from '../../../../utils/Debug.js';
import { getRequestQuery } from '../../../../utils/getRequestQuery.js';

const debug = debugLibrary('searchIDs');

const searchIDs = {
  method: ['GET', 'POST'],
  schema: {
    summary:
      'Retrieve articles which title or abstract contains the given text',
    description: 'Allows to search for articles Title and Abstract.',
    querystring: {
      patentsIDs: {
        type: 'string',
        description: 'patents IDs',
        example: 'EP-2078065-A2, EP-1293521-A2',
        default: '',
      },
      keywords: {
        type: 'string',
        description: 'Text to be searched in articles Title and Abstract',
        example: 'antibiotic',
        default: '',
      },
      fields: {
        type: 'string',
        description: 'Fields to retrieve',
        default: 'data',
      },
      minScore: {
        type: 'number',
        description: 'Text search score minimum',
        example: 2,
        default: 1,
      },
      limit: {
        type: 'number',
        description: 'Maximum number of results to return',
        example: 200,
        default: 100,
      },
    },
  },
  handler: searchHandler,
};

export default searchIDs;

async function searchHandler(request) {
  let data = getRequestQuery(request);
  let {
    patentsIDs = '',
    keywords = '',
    fields = 'data.title, _id, data.abstract',
    minScore = 0,
    limit = 100,
  } = data;
  let formattedFields = getFields(fields);
  let connection;
  try {
    connection = new OctoChemConnection();
    const collection = await connection.getCollection('patents');

    let matchParameters = {};
    let aggregateParameters;
    if (patentsIDs !== '') {
      matchParameters._id = { $in: patentsIDs.split(/[ ,;\t\r\n]+/) };
    }
    if (keywords !== '') {
      matchParameters.$text = { $search: keywords };
      formattedFields.score = { $meta: 'textScore' };
      aggregateParameters = [
        {
          $match: matchParameters,
        },

        { $project: formattedFields },
        {
          $match: { score: { $gte: minScore } },
        },

        { $limit: Number(limit) },
      ];
    }
    if (keywords === '') {
      aggregateParameters = [
        {
          $match: matchParameters,
        },
        { $project: formattedFields },
        { $limit: Number(limit) },
      ];
    }

    const result = await collection.aggregate(aggregateParameters).toArray();

    return { data: result };
  } catch (e) {
    if (connection) {
      await debug.error(e.message, {
        collection: 'patents',
        connection,
        stack: e.stack,
      });
    }
    return { errors: [{ title: e.message, detail: e.stack }] };
  } finally {
    debug.trace('Closing connection');
    if (connection) await connection.close();
  }
}
