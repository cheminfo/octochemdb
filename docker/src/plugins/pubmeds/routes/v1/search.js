// query for molecules from monoisotopic mass
import { OctoChemConnection, getFields } from '../../../../server/utils.js';
import debugLibrary from '../../../../utils/Debug.js';
import { getRequestQuery } from '../../../../utils/getRequestQuery.js';

const debug = debugLibrary('search');

const search = {
  method: ['GET', 'POST'],
  schema: {
    summary:
      'Retrieve articles which title, MeSH terms or abstract contains the given text',
    description: 'Allows to search for articles Title and Abstract.',
    querystring: {
      ids: {
        type: 'string',
        description: 'PubMed IDs comma separated',
        example: '19342308,17200418',
        default: '',
      },
      keywords: {
        type: 'string',
        description:
          'Text to be searched in articles Title, Abstract or MeSH terms',
        example: 'antibiotic',
        default: '',
      },
      fields: {
        type: 'string',
        description: 'Fields to retrieve',
        default: 'data',
      },
      minScore: {
        type: 'number',
        description: 'Text search score minimum',
        example: 2,
        default: 1,
      },
      limit: {
        type: 'number',
        description: 'Maximum number of results to return',
        example: 200,
        default: 100,
      },
    },
  },
  handler: searchHandler,
};

export default search;

async function searchHandler(request) {
  let data = getRequestQuery(request);
  let {
    ids = '',
    keywords = '',
    fields = 'data',
    minScore = 0,
    limit = 100,
  } = data;
  let formattedFields = getFields(fields);
  let connection;
  try {
    connection = new OctoChemConnection();
    const collection = await connection.getCollection('pubmeds');
    let matchParameters = {};
    let aggregateParameters;

    if (keywords !== '') {
      matchParameters.$text = { $search: keywords };
      formattedFields.score = { $meta: 'textScore' };
    }
    if (ids !== '') {
      matchParameters._id = {
        $in: ids
          .split(/[ ,;\t\r\n]+/)
          .filter((entry) => entry)
          .map(Number),
      };
    }
    if (keywords !== '' && ids !== '') {
      aggregateParameters = [
        {
          $match: matchParameters,
        },

        { $project: formattedFields },
        {
          $match: { score: { $gt: minScore } },
        },

        { $limit: Number(limit) },
      ];
    }
    if (keywords === '') {
      aggregateParameters = [
        {
          $match: matchParameters,
        },
        { $project: formattedFields },
        { $limit: Number(limit) },
      ];
    }

    const result = await collection.aggregate(aggregateParameters).toArray();

    return { data: result };
  } catch (e) {
    if (connection) {
      await debug.error(e.message, {
        collection: 'pubmeds',
        connection,
        stack: e.stack,
      });
    }
    return { errors: [{ title: e.message, detail: e.stack }] };
  } finally {
    debug.trace('Closing connection');
    if (connection) await connection.close();
  }
}
