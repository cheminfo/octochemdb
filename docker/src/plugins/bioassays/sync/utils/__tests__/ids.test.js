import { test, expect } from 'vitest';

import { OctoChemConnection } from '../../../../../utils/OctoChemConnection.js';
import search from '../../../routes/v1/ids.js';

test('id search (bioassays)', async () => {
  const connection = new OctoChemConnection();

  while (true) {
    const collection = await connection.getCollection('bioassays');
    if ((await collection.countDocuments()) === 20) {
      break;
    }
  }
  const request = {
    query: {
      ids: '59478_1,5351641_1',
      fields: 'data',
    },
  };
  const results = await search.handler(request);
  // @ts-ignore
  expect(Object.keys(results.data[0].data)).toMatchInlineSnapshot(`
    [
      "cid",
      "aid",
      "assay",
      "ocl",
      "targetTaxonomies",
    ]
  `);
  expect(results.data).toMatchSnapshot();
});
