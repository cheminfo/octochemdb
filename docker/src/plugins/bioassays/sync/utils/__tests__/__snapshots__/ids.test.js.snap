// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`id search (bioassays) 2`] = `
[
  {
    "_id": "5351641_1",
    "data": {
      "aid": 1,
      "assay": "uHTS for Small Molecule Inhibitors of Eukaryotic Translation Initiation",
      "cid": 5351641,
      "ocl": {
        "coordinates": "!BkDKv^?DDiHVdaYL}kDJlPoY{}gn[noDDiHVdaZqB^?D}kFslwoaF_zREQJYxN\\\\KL?Fr\\\\DfQAaLg|{OIKcTodY^s|aMXt|cSu?o\`t|o{xMOTt|cSrMOW~~@",
        "noStereoID": "ekTpA@@@LAEMGLn\\\\dTTRbRfLbteRrRTfbqbtRthdRjZFFfNnAQjjjjjjjfjjjjjijjh@@",
        "noStereoTautomerID": "ekTpA@@@LAEMGLn\\\\dTTRbRfLbteRrRTfbqbtRthdRjZFFfNnAQjjjjjjjfjjjjjijjh@@",
      },
      "targetTaxonomies": [
        {
          "class": "Candidatus Borrarchaeia",
          "family": "Candidatus Borrarchaeaceae",
          "genus": "Candidatus Borrarchaeum",
          "order": "Candidatus Borrarchaeales",
          "phylum": "Candidatus Borrarchaeota",
          "superkingdom": "Archaea",
        },
      ],
    },
  },
  {
    "_id": "59478_1",
    "data": {
      "aid": 1,
      "assay": "uHTS for Small Molecule Inhibitors of Eukaryotic Translation Initiation",
      "cid": 59478,
      "ocl": {
        "coordinates": "!BkDKv^?DDiHVdaYL}kDJlPoY{}gn[noDDiHVdaZqB^?D}kFslwoaF_zREQJYxN\\\\KL?Fr\\\\DfQAaLg|{OIKcTodY^s|aMXt|cSu?o\`t|o{xMOTt|cSrMOW~~@",
        "noStereoID": "ekTpA@@@LAEMGLn\\\\dTTRbRfLbteRrRTfbqbtRthdRjZFFfNnAQjjjjjjjfjjjjjijjh@@",
        "noStereoTautomerID": "ekTpA@@@LAEMGLn\\\\dTTRbRfLbteRrRTfbqbtRthdRjZFFfNnAQjjjjjjjfjjjjjijjh@@",
      },
      "targetTaxonomies": [
        {
          "class": "Candidatus Borrarchaeia",
          "family": "Candidatus Borrarchaeaceae",
          "genus": "Candidatus Borrarchaeum",
          "order": "Candidatus Borrarchaeales",
          "phylum": "Candidatus Borrarchaeota",
          "superkingdom": "Archaea",
        },
      ],
    },
  },
]
`;
